luna_soc/gateware/core/usb2/ep_control.py

```diff
@@ -0,0 +1,175 @@
+#
+# This file is part of LUNA.
+#
+# Copyright (c) 2020-2025 Great Scott Gadgets <info@greatscottgadgets.com>
+# SPDX-License-Identifier: BSD-3-Clause
+
+""" Implementation of a Triple-FIFO endpoint manager.
+
+Equivalent (but not binary-compatbile) implementation of ValentyUSB's ``eptri``.
+
+For an example, see ``examples/usb/eptri`` or TinyUSB's ``luna/dcd_eptri.c``.
+"""
+
+from typing                           import Annotated
+
+from amaranth                         import *
+from amaranth.hdl.xfrm                import ResetInserter, DomainRenamer
+from amaranth.lib                     import wiring
+from amaranth.lib.fifo                import SyncFIFOBuffered
+from amaranth.lib.wiring              import In, Out, connect, flipped
+
+from amaranth_soc                     import csr, event
+
+from luna.gateware.usb.usb2.endpoint  import EndpointInterface
+
+
+class Peripheral(wiring.Component):
+    """ Setup component of our `eptri`-equivalent interface.
+
+    Implements the USB Setup FIFO, which handles SETUP packets on any endpoint.
+
+    This interface is similar to an :class:`OutFIFOInterface`, but always ACKs packets,
+    and does not allow for any flow control; as a USB device must always be ready to accept
+    control packets. [USB2.0: 8.6.1]
+
+    Attributes
+    ----------
+
+    interface: EndpointInterface
+        Our primary interface to the core USB device hardware.
+    """
+
+    class Control(csr.Register, access="w"):
+        """ Control register
+
+            address: Controls the current device's USB address. Should be written after a SET_ADDRESS
+                     request is received. Automatically resets back to zero on a USB reset.
+        """
+        address : csr.Field(csr.action.W,       unsigned(8))
```

Comment:
What's the reason for separating the address RW register in two?

Reply:
So this is pretty cool! With the new register field access model we also got an ownership model that makes explicit the relationship between peripheral and the bus-initiator. So, in this case, you can no longer have `RW` field _and_ have a strobe field under control of the peripheral as this can introduce write conflicts.

Full details: https://amaranth-lang.org/rfcs/0016-soc-csr-regs.html#field-access-and-ownership